### PR TITLE
Fix lots of build warnings

### PR DESCRIPTION
**Pull request by daleglass**
With this the interface build is almost clean.

This adds a warning whitelist system -- there's some stuff we just can't fix yet because we can't upgrade Qt or OpenSSL. This also adds a way to make the build fail on warning, and this is intended to be enabled by default very soon to improve code quality.

### CODE REVIEW

**Comment by daleglass** [comment]
Pulled the contents of PR #77 into this one.

File: CMakeLists.txt [hunk 1]
```diff
@@ -133,6 +133,40 @@ if( NOT WIN32 )
...
+    set(OVERTE_WARNINGS_AS_ERRORS false CACHE BOOL "Count warnings as errors")
+endif()
+
+if(OVERTE_WARNINGS_WHITELIST)
+    if (CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+        message("GCC compiler detected, suppressing some unsolvable warnings.")
+        add_compile_definitions(OVERTE_WARNINGS_WHITELIST_GCC)
+    elseif(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+        message("Clang compiler detected, suppressing some unsolvable warnings.")
+        add_compile_definitions(OVERTE_WARNINGS_WHITELIST_GLANG)
```
Comment (JulianGro) [inline CMakeLists.txt:160]:
```suggestion
        add_compile_definitions(OVERTE_WARNINGS_WHITELIST_CLANG)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -133,6 +133,40 @@ if( NOT WIN32 )
...
+#
+# We can also treat warnings as errors. Without the whitelist this will almost certainly lead
+# to a build failure.
+
+if(NOT DEFINED OVERTE_WARNINGS_WHITELIST)
+    set(OVERTE_WARNINGS_WHITELIST true CACHE BOOL "Whitelist some warnings we can't currently fix")
+endif()
+
+if(NOT DEFINED OVERTE_WARNINGS_AS_ERRORS)
+    #message("Enabling warnings as errors")
```
Comment (JulianGro) [inline CMakeLists.txt:150]: This looks like it shouldn't be here.
```suggestion
```

**Review by JulianGro — CHANGES_REQUESTED** [review]

**Comment by daleglass** [comment]
Okay, this probably needs cleaning here and there, but I think content-wise it's done and ready for review.

There's a few related warnings related to glm that will be addressed later. Also there are some coming from server code that also will be addressed later.

**Review by JulianGro — APPROVED** [review]

**Comment by AnimatedMannequin** [comment]
Here are some builds you may use for testing: 

**self-hosted_debian-11_aarch64, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-82b8743.sh
 - https://public.overte.org/build/overte/PR76/Overte-PR76-82b8743.tar.Z
 - https://public.overte.org/build/overte/PR76/Overte-PR76-82b8743.tar.gz

**macOS-10.15, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-82b8743.dmg

**ubuntu-18.04, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-82b8743.sh
 - https://public.overte.org/build/overte/PR76/Overte-PR76-82b8743.tar.gz
 - https://public.overte.org/build/overte/PR76/Overte-PR76-82b8743.tar.Z

**Comment by JulianGro** [comment]
Windows builds are failing. See: https://github.com/overte-org/overte/runs/6755226576?check_suite_focus=true#step:13:290 and above

**Comment by AnimatedMannequin** [comment]
Here are some builds you may use for testing: 

**ubuntu-18.04, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-63dfe07.tar.gz
 - https://public.overte.org/build/overte/PR76/Overte-PR76-63dfe07.tar.Z
 - https://public.overte.org/build/overte/PR76/Overte-PR76-63dfe07.sh

**self-hosted_debian-11_aarch64, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-63dfe07.tar.gz
 - https://public.overte.org/build/overte/PR76/Overte-PR76-63dfe07.sh
 - https://public.overte.org/build/overte/PR76/Overte-PR76-63dfe07.tar.Z

**macOS-10.15, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-63dfe07.dmg

**Comment by AnimatedMannequin** [comment]
Here are some builds you may use for testing: 

**self-hosted_debian-11_aarch64, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-5fa7b49.sh
 - https://public.overte.org/build/overte/PR76/Overte-PR76-5fa7b49.tar.Z
 - https://public.overte.org/build/overte/PR76/Overte-PR76-5fa7b49.tar.gz

**macOS-10.15, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-5fa7b49.dmg

**ubuntu-18.04, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-5fa7b49.tar.Z
 - https://public.overte.org/build/overte/PR76/Overte-PR76-5fa7b49.tar.gz
 - https://public.overte.org/build/overte/PR76/Overte-PR76-5fa7b49.sh

**Comment by AnimatedMannequin** [comment]
Here are some builds you may use for testing: 

**self-hosted_debian-11_aarch64, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-e858cea.sh
 - https://public.overte.org/build/overte/PR76/Overte-PR76-e858cea.tar.gz
 - https://public.overte.org/build/overte/PR76/Overte-PR76-e858cea.tar.Z

**ubuntu-18.04, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-e858cea.sh
 - https://public.overte.org/build/overte/PR76/Overte-PR76-e858cea.tar.gz
 - https://public.overte.org/build/overte/PR76/Overte-PR76-e858cea.tar.Z

**Comment by daleglass** [comment]
Okay, I think I finally figured out what's the issue with Windows builds.

We're not actually doing things right in CMake here, we're trying to detect the compiler too early on Windows, and it's returning an empty string.

For now I added an ugly hack that should work. Later I'll have to do a proper PR to deal with this. It'll be a separate PR as it's also going to affect other code I used as inspiration for this work.

**Comment by AnimatedMannequin** [comment]
Here are some builds you may use for testing: 

**ubuntu-18.04, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-2fa623d.tar.Z
 - https://public.overte.org/build/overte/PR76/Overte-PR76-2fa623d.tar.gz
 - https://public.overte.org/build/overte/PR76/Overte-PR76-2fa623d.sh

**self-hosted_debian-11_aarch64, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-2fa623d.tar.Z
 - https://public.overte.org/build/overte/PR76/Overte-PR76-2fa623d.sh
 - https://public.overte.org/build/overte/PR76/Overte-PR76-2fa623d.tar.gz

**macOS-10.15, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-2fa623d.dmg

**Comment by AnimatedMannequin** [comment]
Here are some builds you may use for testing: 

**self-hosted_debian-11_aarch64, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-35fc664.tar.Z
 - https://public.overte.org/build/overte/PR76/Overte-PR76-35fc664.sh
 - https://public.overte.org/build/overte/PR76/Overte-PR76-35fc664.tar.gz

**macOS-10.15, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-35fc664.dmg

**ubuntu-18.04, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-35fc664.sh
 - https://public.overte.org/build/overte/PR76/Overte-PR76-35fc664.tar.gz
 - https://public.overte.org/build/overte/PR76/Overte-PR76-35fc664.tar.Z

**windows-2019, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-35fc664.exe

File: libraries/octree/src/OctreeElement.cpp [hunk 3]
```diff
@@ -443,10 +443,13 @@ void OctreeElement::printDebugDetails(const char* label) const {
...
-                         (double)_cube.getScale(),
-                         debug::valueOf(isLeaf()), debug::valueOf(isDirty()), debug::valueOf(getShouldRender()));
+    qCDebug(octree).noquote() << label
+                              << QString(" - Voxel at corner=(%1,%2,%3)").arg((double)_cube.getCorner().x, (double)_cube.getCorner().y, (double)_cube.getCorner().z)
+                              << "size=" << (double)_cube.getScale()
+                              << " isLeaf=" << debug::valueOf(isLeaf())
+                              << " isDirty=" << debug::valueOf(isDirty())
+                              << " shouldRender=" << debug::valueOf(getShouldRender());
+
     qCDebug(octree).nospace() << resultString;
```
Comment (ksuprynowicz) [inline libraries/octree/src/OctreeElement.cpp:453]: Why is empty string sent to qCDebug(octree).nospace()?

**Comment by daleglass** [comment]
I saw some suspicious looking timeouts when testing this earlier, I might have messed up the QTimer related changes.

**Comment by AnimatedMannequin** [comment]
Here are some builds you may use for testing: 

**macOS-10.15, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-cdd2f9a.dmg

**self-hosted_debian-11_aarch64, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-cdd2f9a.sh
 - https://public.overte.org/build/overte/PR76/Overte-PR76-cdd2f9a.tar.Z
 - https://public.overte.org/build/overte/PR76/Overte-PR76-cdd2f9a.tar.gz

**ubuntu-18.04, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-cdd2f9a.tar.gz
 - https://public.overte.org/build/overte/PR76/Overte-PR76-cdd2f9a.sh
 - https://public.overte.org/build/overte/PR76/Overte-PR76-cdd2f9a.tar.Z

**windows-2019, full**
 - https://public.overte.org/build/overte/PR76/Overte-PR76-cdd2f9a.exe

**Comment by daleglass** [comment]
Spectator camera isn't working. 

`[06/11 21:14:17] [WARNING] [hifi.qml] Unauthorized QML URL found QUrl("https://more.overte.org/applications/spectator-camera/SpectatorCamera.qml")`

This may or not be related to this PR, needs testing.

**Comment by daleglass** [comment]
Settings/QML whitelist is broken

**Comment by daleglass** [comment]
Server tested, everything seems to work.

**Pull request merged**